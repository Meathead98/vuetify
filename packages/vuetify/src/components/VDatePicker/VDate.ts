// Mixins
import Localable from '../../mixins/localable'

// Utils
import { pad, createNativeLocaleFormatter } from './util'
import isDateAllowed, { AllowedDateFunction } from './util/isDateAllowed'
import { consoleWarn } from '../../util/console'
import mixins from '../../util/mixins'

// Types
import { PropValidator } from 'vue/types/options'
import { DatePickerFormatter } from './util/createNativeLocaleFormatter'
import { VNode } from 'vue'
import { wrapInArray } from '../../util/helpers'

export type DateEventColorValue = string | string[]
export type DateEvents = string[] | ((date: string) => boolean | DateEventColorValue) | Record<string, DateEventColorValue>
export type DateEventColors = DateEventColorValue | Record<string, DateEventColorValue> | ((date: string) => DateEventColorValue)
export type DatePickerMultipleFormatter = (date: string[]) => string

export interface VDateFormatters {
  year: DatePickerFormatter
  titleDate: DatePickerFormatter | DatePickerMultipleFormatter
  landscapeTitleDate: DatePickerFormatter | DatePickerMultipleFormatter
  headerDate: DatePickerFormatter
  headerMonth: DatePickerFormatter
  date: DatePickerFormatter
  month: DatePickerFormatter
  weekday: DatePickerFormatter
}

// Adds leading zero to month/day if necessary, returns 'YYYY' if type = 'year',
// 'YYYY-MM' if 'month' and 'YYYY-MM-DD' if 'date'
function sanitizeDateString (dateString: string, type: 'date' | 'month' | 'year'): string {
  const [year, month = 1, date = 1] = dateString.split('-')
  return `${year}-${pad(month)}-${pad(date)}`.substr(0, { date: 10, month: 7, year: 4 }[type])
}

export const enum PickerType {
  Year = 'year',
  Month = 'month',
  Date = 'date'
}

type DatePickerValue = string | string[] | undefined
type DatePickerType = 'date' | 'month' | 'year'

export interface VDateScopedProps {
  currentDate: String
  dateClick: Function
  monthClick: Function
  yearClick: Function
  formatters: VDateFormatters
  value: string[]
  activePicker: PickerType
  updateActivePicker: Function
  pickerDate: string
  updatePickerDate: Function
  multiple: boolean
}

export const VDateProps = {
  formatters: {
    type: Object,
    default: () => ({})
  } as PropValidator<Partial<VDateFormatters>>,
  activePicker: String as PropValidator<DatePickerType>,
  allowedDates: Function as PropValidator<AllowedDateFunction | undefined>,
  currentDate: String,
  max: String,
  min: String,
  multiple: Boolean,
  pickerDate: String,
  reactive: Boolean,
  value: [Array, String] as PropValidator<DatePickerValue>,
  type: {
    type: String,
    default: 'date',
    validator: (type: any) => ['month', 'date'].includes(type) // TODO: add year
  } as any as PropValidator<DatePickerType>
}

export default mixins(
  Localable
/* @vue/component */
).extend({
  name: 'v-date',

  inheritAttrs: false,

  props: {
    ...VDateProps
  },

  data () {
    const firstDate = this.value && wrapInArray(this.value)[0]
    const now = this.currentDate || new Date().toISOString()
    const nowDate = sanitizeDateString(now, 'date')
    const internalPickerDate = sanitizeDateString(this.pickerDate || firstDate || nowDate, this.type === 'month' ? 'year' : 'month')

    return {
      now: nowDate,
      internalPickerDate,
      internalDate: this.value ? wrapInArray(this.value).map(date => sanitizeDateString(date, this.type)) : [],
      internalActivePicker: (this.activePicker || this.type) as PickerType
    }
  },

  computed: {
    lastPickedValue (): string {
      return this.internalDate[this.internalDate.length - 1]
    },
    computedFormatters (): VDateFormatters {
      return {
        year: this.formatters.year || createNativeLocaleFormatter(this.currentLocale, { year: 'numeric', timeZone: 'UTC' }, { length: 4 }),
        titleDate: this.formatters.titleDate || this.defaultTitleDateFormatter,
        landscapeTitleDate: this.formatters.landscapeTitleDate || this.defaultLandscapeTitleDate,
        weekday: this.formatters.weekday || createNativeLocaleFormatter(this.currentLocale, { weekday: 'narrow', timeZone: 'UTC' }) || (v => v),
        date: this.formatters.date || createNativeLocaleFormatter(this.currentLocale, { day: 'numeric', timeZone: 'UTC' }, { start: 8, length: 2 }) || (v => v),
        headerMonth: this.formatters.headerMonth || createNativeLocaleFormatter(this.currentLocale, { month: 'long', year: 'numeric', timeZone: 'UTC' }, { length: 7 }),
        month: this.formatters.month || createNativeLocaleFormatter(this.currentLocale, { month: 'short', timeZone: 'UTC' }, { start: 5, length: 2 }),
        headerDate: this.formatters.headerDate || createNativeLocaleFormatter(this.currentLocale, { year: 'numeric', timeZone: 'UTC' }, { length: 4 })
      }
    },
    defaultTitleDateFormatter (): DatePickerMultipleFormatter {
      const titleFormats = {
        year: { year: 'numeric', timeZone: 'UTC' },
        month: { month: 'long', timeZone: 'UTC' },
        date: { weekday: 'short', month: 'short', day: 'numeric', timeZone: 'UTC' }
      }

      const substrOptions = { date: 10, month: 7, year: 4 }

      const titleDateFormatter = createNativeLocaleFormatter(this.currentLocale, titleFormats[this.type], {
        start: 0,
        length: substrOptions[this.type]
      })

      if (this.multiple) {
        if (this.internalDate.length < 2) return dates => dates.length ? titleDateFormatter(dates[0]) : '0 selected' // TODO: localization
        else return dates => `${dates.length} selected` // TODO: localization
      } else {
        return dates => dates.length ? titleDateFormatter(dates[0]) : '-'
      }
    },
    defaultLandscapeTitleDate (): DatePickerMultipleFormatter {
      return (date: string[]) => this.defaultTitleDateFormatter(date)
        .replace(/([^\d\s])([\d])/g, (match, nonDigit, digit) => `${nonDigit} ${digit}`)
        .replace(', ', ',<br>')
    },
    scopedSlotProps (): VDateScopedProps {
      return {
        currentDate: this.now,
        dateClick: this.dateClick,
        monthClick: this.monthClick,
        yearClick: this.yearClick,
        formatters: this.computedFormatters,
        value: this.internalDate,
        activePicker: this.internalActivePicker,
        updateActivePicker: this.updateActivePicker,
        pickerDate: this.internalPickerDate,
        updatePickerDate: this.updatePickerDate,
        multiple: this.multiple
      }
    }
  },

  watch: {
    internalPickerDate (val: string, prev: string) {
      // TODO: What was isReversing used for??
      // Make a ISO 8601 strings from val and prev for comparision, otherwise it will incorrectly
      // compare for example '2000-9' and '2000-10'
      // const sanitizeType = this.type === 'month' ? 'year' : 'month'
      // this.isReversing = sanitizeDateString(val, sanitizeType) < sanitizeDateString(prev, sanitizeType)
      this.$emit('update:pickerDate', val)
    },
    pickerDate (val: string | null) {
      if (val) {
        this.internalPickerDate = val
      } else if (this.lastPickedValue) {
        this.internalPickerDate = sanitizeDateString(this.lastPickedValue, 'month')
      }
    },
    value (newValue: DatePickerValue, oldValue: DatePickerValue) {
      this.checkMultipleProp()

      if (newValue) {
        const arr = wrapInArray(newValue)

        if (!arr.every(this.isDateAllowed)) return

        this.internalDate = arr

        if (!this.pickerDate) {
          this.internalPickerDate = sanitizeDateString(this.lastPickedValue, this.type === 'month' ? 'year' : 'month')
        }
      }
    },
    type (type: PickerType) {
      this.internalActivePicker = type

      // If we're switching type then we need to reformat current values
      // e.g. 2019-05-01 => 2019-05 when going from date to month
      if (this.internalDate.length) {
        this.internalDate = this.internalDate.map(date => sanitizeDateString(date, type)).filter(this.isDateAllowed)
      }
    },
    internalDate (internalDate: string[]) {
      this.emitInput()
    },
    activePicker (activePicker: DatePickerType) {
      this.internalActivePicker = activePicker as PickerType
    }
  },

  created () {
    this.checkMultipleProp()
  },

  methods: {
    updateActivePicker (type: PickerType) {
      this.internalActivePicker = type
      this.$emit('update:activePicker', type)
    },
    updatePickerDate (date: string) {
      this.internalPickerDate = date
    },
    emitInput () {
      if (!this.internalDate.length) return

      const value = this.multiple ? this.internalDate : this.internalDate[0]

      this.$emit('input', value)
      !this.multiple && this.$emit('change', value)
    },
    checkMultipleProp () {
      if (this.value == null) return
      const valueType = this.value.constructor.name
      const expected = this.multiple ? 'Array' : 'String'
      if (valueType !== expected) {
        consoleWarn(`Value must be ${this.multiple ? 'an' : 'a'} ${expected}, got ${valueType}`, this)
      }
    },
    isDateAllowed (value: string): boolean {
      return isDateAllowed(value, this.min, this.max, this.allowedDates)
    },
    yearClick (value: string) {
      if (this.type !== 'year') {
        this.internalPickerDate = value
        this.internalActivePicker = PickerType.Month
      } else {
        this.updateInternalDate(value)
      }
    },
    monthClick (value: string) {
      if (this.type === 'date') {
        this.internalPickerDate = value
        this.internalActivePicker = PickerType.Date
      } else {
        this.updateInternalDate(value)
      }
    },
    dateClick (value: string) {
      this.updateInternalDate(value)
    },
    updateInternalDate (value: string) {
      if (this.internalDate.includes(value)) {
        this.internalDate = this.internalDate.filter(d => d !== value)
      } else if (this.multiple) {
        this.internalDate.push(value)
        this.internalDate.sort()
      } else {
        this.internalDate = [value]
      }
    }
  },

  render (): VNode {
    return this.$scopedSlots.default!(this.scopedSlotProps) as any
  }
})
