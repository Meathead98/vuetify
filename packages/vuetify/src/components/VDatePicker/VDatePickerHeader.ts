import './VDatePickerHeader.sass'

// Components
import VBtn from '../VBtn'
import VIcon from '../VIcon'

// Mixins
import Colorable from '../../mixins/colorable'
import Localable from '../../mixins/localable'
import Themeable from '../../mixins/themeable'

// Utils
import { monthChange } from './util'
import mixins from '../../util/mixins'

// Types
import { VNode } from 'vue'
import { DatePickerFormatter } from './util/createNativeLocaleFormatter'
import { PropValidator } from 'vue/types/options'
import { PickerType } from './VDate'

export default mixins(
  Colorable,
  Localable,
  Themeable
/* @vue/component */
).extend({
  name: 'v-date-picker-header',

  inheritAttrs: false,

  props: {
    disabled: Boolean,
    yearFormat: Function as PropValidator<DatePickerFormatter>,
    monthFormat: Function as PropValidator<DatePickerFormatter>,
    min: String,
    max: String,
    nextIcon: {
      type: String,
      default: '$vuetify.icons.next'
    },
    prevIcon: {
      type: String,
      default: '$vuetify.icons.prev'
    },
    readonly: Boolean,
    /** value is string formatted as year-month (e.g. 2005-11) */
    value: {
      type: String,
      required: true
    },
    activePicker: String as PropValidator<PickerType>
  },

  data () {
    return {
      isReversing: false
    }
  },

  computed: {
    isMonthPicker (): boolean {
      return this.activePicker === PickerType.Month
    },
    isDatePicker (): boolean {
      return this.activePicker === PickerType.Date
    },
    formatter (): DatePickerFormatter {
      const fn = (v: string) => v
      switch (this.activePicker) {
        case PickerType.Date: return this.monthFormat || fn
        case PickerType.Month: return this.yearFormat || fn
        default: return fn
      }
    }
  },

  watch: {
    value (newVal, oldVal) {
      this.isReversing = newVal < oldVal
    }
  },

  methods: {
    genBtn (change: number) {
      const newDate = this.calculateChange(this.value, change, this.activePicker)
      const disabled = this.disabled ||
        (change < 0 && this.min && newDate < this.min) ||
        (change > 0 && this.max && newDate > this.max)

      return this.$createElement(VBtn, {
        props: {
          dark: this.dark,
          disabled,
          icon: true,
          light: this.light
        },
        nativeOn: {
          click: (e: Event) => {
            e.stopPropagation()
            this.$emit(`update:pickerDate`, newDate)
          }
        }
      }, [
        this.$createElement(VIcon, ((change < 0) === !this.$vuetify.rtl) ? this.prevIcon : this.nextIcon)
      ])
    },
    calculateChange (value: string, sign: number, activePicker: PickerType) {
      const [year] = value.split('-').map(Number)

      switch (activePicker) {
        case PickerType.Date: return monthChange(value, sign)
        case PickerType.Month: return `${year + sign}`
        default: return value
      }
    },
    genHeader () {
      const color = !this.disabled && (this.color || 'accent')
      const header = this.$createElement('div', this.setTextColor(color, {
        key: this.value
      }), [this.$createElement('button', {
        attrs: {
          type: 'button'
        },
        on: {
          click: () => this.$emit('update:activePicker', this.isDatePicker ? PickerType.Month : PickerType.Year)
        }
      }, [this.$slots.default || this.formatter(this.value)])])

      const transition = this.$createElement('transition', {
        props: {
          name: (this.isReversing === !this.$vuetify.rtl) ? 'tab-reverse-transition' : 'tab-transition'
        }
      }, [header])

      return this.$createElement('div', {
        staticClass: 'v-date-picker-header__value',
        class: {
          'v-date-picker-header__value--disabled': this.disabled
        }
      }, [transition])
    }
  },

  render (): VNode {
    return this.$createElement('div', {
      staticClass: 'v-date-picker-header',
      class: {
        'v-date-picker-header--disabled': this.disabled,
        ...this.themeClasses
      }
    }, [
      this.genBtn(-1),
      this.genHeader(),
      this.genBtn(+1)
    ])
  }
})
