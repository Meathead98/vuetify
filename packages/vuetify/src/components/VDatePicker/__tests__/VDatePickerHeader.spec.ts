import VDatePickerHeader from '../VDatePickerHeader'
import { Lang } from '../../../services/lang'
import {
  mount,
  MountOptions,
  Wrapper
} from '@vue/test-utils'
import Vue from 'vue'
import { PickerType } from '../VDate'

Vue.prototype.$vuetify = {
  icons: {
    values: {
      next: 'mdi-chevron-right',
      prev: 'mdi-chevron-left'
    }
  }
}

describe('VDatePickerHeader.ts', () => {
  type Instance = InstanceType<typeof VDatePickerHeader>
  let mountFunction: (options?: MountOptions<Instance>) => Wrapper<Instance>

  beforeEach(() => {
    mountFunction = (options?: MountOptions<Instance>) => {
      return mount(VDatePickerHeader, {
        ...options,
        mocks: {
          $vuetify: {
            rtl: false,
            lang: new Lang()
          }
        }
      })
    }
  })

  it('should render component and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        value: '2005-11'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render disabled component and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        value: '2005-11',
        disabled: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render readonly component and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        value: '2005-11',
        readonly: true
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component in RTL mode and match snapshot', async () => {
    const wrapper = mountFunction({
      propsData: {
        value: '2005-11'
      }
    })

    ;(wrapper.vm as any).$vuetify.rtl = true
    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    ;(wrapper.vm as any).$vuetify.rtl = undefined
  })

  it('should render prev/next icons', () => {
    const wrapper = mountFunction({
      propsData: {
        value: '2005',
        prevIcon: 'foo',
        nextIcon: 'bar'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with own formatter and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        activePicker: PickerType.Date,
        value: '2005-11',
        monthFormat: value => `(${value})`,
        yearFormat: value => `<${value}>`
      }
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      activePicker: PickerType.Month
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render colored component and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        value: '2005-11',
        color: 'green lighten-1'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with default slot and match snapshot', () => {
    const wrapper = mountFunction({
      propsData: {
        value: '2005-11'
      },
      slots: {
        default: '<span>foo</span>'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should trigger activePicker event when clicking on header', () => {
    const activePicker = jest.fn()
    const wrapper = mountFunction({
      propsData: {
        activePicker: PickerType.Date,
        value: '2005-11'
      },
      listeners: {
        'update:activePicker': activePicker
      }
    })

    wrapper.find('.v-date-picker-header__value button').trigger('click')
    expect(activePicker).toHaveBeenCalledWith(PickerType.Month)
  })

  it('should trigger update events on arrows click', () => {
    const pickerDate = jest.fn()
    const wrapper = mountFunction({
      propsData: {
        activePicker: PickerType.Date,
        value: '2005-12'
      },
      listeners: {
        'update:pickerDate': pickerDate
      }
    })

    wrapper.findAll('button.v-btn').at(0).trigger('click')
    expect(pickerDate).toHaveBeenCalledWith('2005-11')

    wrapper.findAll('button.v-btn').at(1).trigger('click')
    expect(pickerDate).toHaveBeenCalledWith('2006-01')

    wrapper.setProps({
      activePicker: PickerType.Month
    })

    wrapper.findAll('button.v-btn').at(0).trigger('click')
    expect(pickerDate).toHaveBeenCalledWith('2004')

    wrapper.findAll('button.v-btn').at(1).trigger('click')
    expect(pickerDate).toHaveBeenCalledWith('2006')
  })
})
