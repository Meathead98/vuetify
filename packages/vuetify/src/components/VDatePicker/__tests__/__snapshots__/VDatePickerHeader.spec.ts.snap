// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VDatePickerHeader.ts should render colored component and match snapshot 1`] = `

<div class="v-date-picker-header theme--light">
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-left theme--light"
      >
      </i>
    </span>
  </button>
  <div class="v-date-picker-header__value">
    <div class="green--text text--lighten-1">
      <button type="button">
        2005-11
      </button>
    </div>
  </div>
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-right theme--light"
      >
      </i>
    </span>
  </button>
</div>

`;

exports[`VDatePickerHeader.ts should render component and match snapshot 1`] = `

<div class="v-date-picker-header theme--light">
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-left theme--light"
      >
      </i>
    </span>
  </button>
  <div class="v-date-picker-header__value">
    <div class="accent--text">
      <button type="button">
        2005-11
      </button>
    </div>
  </div>
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-right theme--light"
      >
      </i>
    </span>
  </button>
</div>

`;

exports[`VDatePickerHeader.ts should render component in RTL mode and match snapshot 1`] = `

<div class="v-date-picker-header theme--light">
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-right theme--light"
      >
      </i>
    </span>
  </button>
  <div class="v-date-picker-header__value">
    <div class="accent--text">
      <button type="button">
        2005-11
      </button>
    </div>
  </div>
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-left theme--light"
      >
      </i>
    </span>
  </button>
</div>

`;

exports[`VDatePickerHeader.ts should render component with default slot and match snapshot 1`] = `

<div class="v-date-picker-header theme--light">
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-left theme--light"
      >
      </i>
    </span>
  </button>
  <div class="v-date-picker-header__value">
    <div class="accent--text">
      <button type="button">
        <span>
          foo
        </span>
      </button>
    </div>
  </div>
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-right theme--light"
      >
      </i>
    </span>
  </button>
</div>

`;

exports[`VDatePickerHeader.ts should render component with own formatter and match snapshot 1`] = `

<div class="v-date-picker-header theme--light">
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-left theme--light"
      >
      </i>
    </span>
  </button>
  <div class="v-date-picker-header__value">
    <div class="accent--text">
      <button type="button">
        (2005-11)
      </button>
    </div>
  </div>
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-right theme--light"
      >
      </i>
    </span>
  </button>
</div>

`;

exports[`VDatePickerHeader.ts should render component with own formatter and match snapshot 2`] = `

<div class="v-date-picker-header theme--light">
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-left theme--light"
      >
      </i>
    </span>
  </button>
  <div class="v-date-picker-header__value">
    <div class="accent--text">
      <button type="button">
        &lt;2005-11&gt;
      </button>
    </div>
  </div>
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-right theme--light"
      >
      </i>
    </span>
  </button>
</div>

`;

exports[`VDatePickerHeader.ts should render disabled component and match snapshot 1`] = `

<div class="v-date-picker-header v-date-picker-header--disabled theme--light">
  <button disabled="disabled"
          type="button"
          class="v-btn v-btn--disabled v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-left theme--light"
      >
      </i>
    </span>
  </button>
  <div class="v-date-picker-header__value v-date-picker-header__value--disabled">
    <div>
      <button type="button">
        2005-11
      </button>
    </div>
  </div>
  <button disabled="disabled"
          type="button"
          class="v-btn v-btn--disabled v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-right theme--light"
      >
      </i>
    </span>
  </button>
</div>

`;

exports[`VDatePickerHeader.ts should render prev/next icons 1`] = `

<div class="v-date-picker-header theme--light">
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon material-icons theme--light"
      >
        foo
      </i>
    </span>
  </button>
  <div class="v-date-picker-header__value">
    <div class="accent--text">
      <button type="button">
        2005
      </button>
    </div>
  </div>
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon material-icons theme--light"
      >
        bar
      </i>
    </span>
  </button>
</div>

`;

exports[`VDatePickerHeader.ts should render readonly component and match snapshot 1`] = `

<div class="v-date-picker-header theme--light">
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-left theme--light"
      >
      </i>
    </span>
  </button>
  <div class="v-date-picker-header__value">
    <div class="accent--text">
      <button type="button">
        2005-11
      </button>
    </div>
  </div>
  <button type="button"
          class="v-btn v-btn--fab v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"
  >
    <span class="v-btn__content">
      <i aria-hidden="true"
         class="v-icon mdi mdi-chevron-right theme--light"
      >
      </i>
    </span>
  </button>
</div>

`;
